server/pulp/server/controllers/distributor.py

```diff
@@ -1,109 +1,294 @@
+    dist_instance.distributor_removed(repo.to_transfer_repo(), call_config)
+    distributor.delete()
 
     unbind_errors = []
     additional_tasks = []
     options = {}
-    manager = managers.consumer_bind_manager()
-    for bind in manager.find_by_distributor(repo_id, distributor_id):
+
+    bind_manager = managers.consumer_bind_manager()
+    for bind in bind_manager.find_by_distributor(repo_id, distributor_id):
         try:
-            report = consumer.unbind(bind['consumer_id'],
-                                     bind['repo_id'],
-                                     bind['distributor_id'],
-                                     options)
+            report = bind_manager.unbind(bind['consumer_id'], bind['repo_id'],
+                                         bind['distributor_id'], options)
             if report:
                 additional_tasks.extend(report.spawned_tasks)
         except Exception, e:
             unbind_errors.append(e)
 
     bind_error = None
-    if len(unbind_errors) > 0:
-        bind_error = PulpCodedException(PLP0003, repo_id=repo_id, distributor_id=distributor_id)
+    if unbind_errors:
+        bind_error = exceptions.PulpCodedException(PLP0003, repo_id=repo_id,
+                                                   distributor_id=distributor_id)
         bind_error.child_exceptions = unbind_errors
+
     return TaskResult(error=bind_error, spawned_tasks=additional_tasks)
 
 
+def queue_update(distributor, config, delta):
+    """
+    Dispatch a task to update a distributor.
+
+    :param distributor: distributor to be updated
+    :type  distributor: pulp.server.db.model.Distributor
+    :param config: A configuration dictionary for a distributor instance. The contents of this dict
+                   depends on the type of distributor. Values of None will remove they key from the
+                   config. Keys ommited from this dictionary will remain unchanged.
+    :type  config: dict
+    :param delta: A dictionary used to change conf values for a distributor instance. This currently
+                  only supports the 'auto_publish' keyword, which should have a value of type bool
+    :type  delta: dict or None
+
+    :return: An AsyncResult instance as returned by Celery's apply_async
+    :rtype:  celery.result.AsyncResult
+    """
+    task_tags = [
+        tags.resource_tag(tags.RESOURCE_REPOSITORY_TYPE, distributor.repo_id),
+        tags.resource_tag(tags.RESOURCE_REPOSITORY_DISTRIBUTOR_TYPE, distributor.distributor_id),
+        tags.action_tag('update_distributor')
+    ]
+    async_result = update.apply_async_with_reservation(
+        tags.RESOURCE_REPOSITORY_TYPE, distributor.repo_id,
+        [distributor, config, delta],
+        tags=task_tags)
+    return async_result
+
+
 @celery.task(base=Task, name='pulp.server.tasks.repository.distributor_update')
-def update(repo_id, distributor_id, config, delta):
+def update(distributor, config, delta):
```

Comment:
For tasks, we are much better off passing IDs than objects. Otherwise the object has to be serialized and stuffed into an AMQP message, which is usually doable, but not the best approach. @shubham90 will thank you.